### PR TITLE
Use elm-tooling-cli to install third party bins

### PR DESCRIPTION
**Pull request by Razzeee**
This already seems pretty good to me, probably still some corner cases left.

- [ ] we can't install a test runner like this, yet - will have to wait for `elm-test-rs`
- [x] can we just call install every time? should we check if we can already find the bin and skip or check if the versions matches?
- [x] update readme
- [x] error when download fails

### CODE REVIEW

**Comment by Razzeee** [comment]
This should basically allow us to get rid of the prerequisites in https://github.com/elm-tooling/elm-language-server#requirements so that new users only need to install the extension

File: src/elmWorkspace.ts [hunk 1]
```diff
@@ -363,6 +352,85 @@ export class ElmWorkspace implements IElmWorkspace {
...
+        "elm-tooling.json",
+      );
+      const options = {
+        cwd: this.rootPath.fsPath,
+        env: process.env,
+        stdin: process.stdin,
+        stderr: process.stderr,
+        stdout: process.stdout,
+      };
+      if (fs.existsSync(pathToElmToolingJson)) {
```
Comment (jmbockhorst) [inline src/elmWorkspace.ts:367]: We should add a `fileExists` method to `IProgramHost` to use here and move this down to `createProgramHost`.

File: src/elmWorkspace.ts [hunk 2]
```diff
@@ -363,6 +352,85 @@ export class ElmWorkspace implements IElmWorkspace {
...
+          await runElmToolingCliInstall(options);
+        } catch (error) {
+          this.connection.window.showErrorMessage(`${error}`);
+          this.connection.console.error(`Unexpected error ${error}`);
+        }
+      } else {
+        this.connection.console.info(`elm-tooling.json does not exist`);
+        this.connection.window
+          .showInformationMessage(
+            "Create elm-tooling.json, to store elm tool versions and download them?",
```
Comment (jmbockhorst) [inline src/elmWorkspace.ts:379]: I'm not sure about this. We are basically forcing users to create an elm-tooling.json, even if they already have their environment setup how they like and working.

File: src/elmWorkspace.ts [hunk 3]
```diff
@@ -363,6 +352,85 @@ export class ElmWorkspace implements IElmWorkspace {
...
+              await this.connection.window.showDocument({
+                uri: pathToElmToolingJson,
+              });
+
+              // we might want to restart, as the used packages might not match the elm version
+              // or we had no elm before and probably were crashing
+              await this.connection.window
+                .createWorkDoneProgress()
+                .then((progress) => {
+                  progress.begin("Restarting Elm Language Server", 0);
```
Comment (jmbockhorst) [inline src/elmWorkspace.ts:402]: Should make a `restart` method that so this is duplicated from above.

**Review by jmbockhorst — COMMENTED** [review]
Maybe we could only offer to create a elm-tooling.json if we can't get the elm version?

 I don't know, I'm split on this. On one hand, it would add visibility to the tool and more people would end up using it, and it's just a notification (I get plenty of bothersome notifications in vscode). On the other hand, we don't want users to feel forced to use it. 

If we do always ask, saying "No" to create elm-tooling.json should set `useElmToolingJsonForTools: true`, so we don't ask again.

File: src/elmWorkspace.ts (hunk 3 above)
Comment (Razzeee) [reply src/elmWorkspace.ts:402]: I did try, but the above was using it via then, which didn't work for me here, as I need it to block I guess? (might have been the `void`)

File: src/elmWorkspace.ts (hunk 2 above)
Comment (Razzeee) [reply src/elmWorkspace.ts:379]: Yeah, it's very opinionated, but that's basically what I'm after. But all I wanted to achive is to cut down on setup, which I'm not even sure this ends up doing in the end. I think relying on node now is the real problem for me here.

**Comment by Razzeee** [comment]
> Maybe we could only offer to create a elm-tooling.json if we can't get the elm version?
> 
> I don't know, I'm split on this. On one hand, it would add visibility to the tool and more people would end up using it, and it's just a notification (I get plenty of bothersome notifications in vscode). On the other hand, we don't want users to feel forced to use it.
> 
> If we do always ask, saying "No" to create elm-tooling.json should set `useElmToolingJsonForTools: true`, so we don't ask again.

I guess that would be sensible and maybe have a flag to force it to use the cli.

I couldn't find a way to write to the settings in the language server api, but maybe I overlooked it.
